lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -233,37 +234,42 @@ static FuncDecl *findOperatorDeclInProtocol(DeclName operatorName,
   return nullptr;
 }
 
-// Return the expected generic signature for autodiff associated functions given
-// a SILDifferentiableAttr. The expected generic signature is built from the
-// original generic signature and the attribute's requirements.
-static CanGenericSignature
-getAssociatedFunctionGenericSignature(SILDifferentiableAttr *attr,
-                                      SILFunction *original) {
+// Return the canonical derivative generic signature for the given original
+// function and (possibly uncanonical) derivative generic signature.
+// The canonical derivative generic signature constrains all wrt parameters
+// to conform to `Differentiable`.
+// TODO(TF-818): Change `@differentiable` attribute type-checking and
+// `[differentiable]` construction so that all derivative generic signatures
+// constrain wrt parameters to `Differentiable`. Then, this helper will no
+// longer be needed (except for constructing attributes in
+// `ADContext::getOrCreateDifferentiableAttr`), improving compiler
+// performance.
+static CanGenericSignature getCanonicalDerivativeGenericSignature(
```

Comment:
Note: `getCanonicalDerivativeGenericSignature` has five users. Caching the results in `ADContext` (based on original function, derivative generic signature, and wrt parameter index subset) would improve performance.

But a more robust fix would be to canonicalize all derivative generic signatures in `DifferentiableAttr` and `SILDifferentiableAttr` so that this canonicalization helper in the differentiation transform is no longer needed. [TF-818](https://bugs.swift.org/browse/TF-818) tracks this.